{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook I will demonstrate by genome assembly algorithm to assemble the genome of *Carsonella ruddii* that has an amazingly small genome of only around 160,000 basepairs. Because genome assembly is an incredibly difficult task that even state-of-the-art algorithms cannot solve with the current read lengths generated by sequencing machines, I will simplify the assembly to only contain error-free paired-end reads.\n",
    "\n",
    "The approach to assemble reads is to generate a de Bruijn graph from the reads. In a de Bruijn graph, reads are broken into short kmers which represent the edges of a graph and the nodes are the (k-1)-mers that connect the kmers. From this, an Eulerian path needs to be walked that visits each edge exactly once to reconstruct the genome.\n",
    "\n",
    "![example](de_bruijn_graph.png)\n",
    "\n",
    "In this example I will use paired reads, which are short reads that are separated by a known distance. This is used by many sequencing protocols to make assembly of shorter reads easier. I will start with the complete genome of *Carsonella ruddii* and first chop this off into paired reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all k-mers of a sequence\n",
    "def get_kmers(input, k):\n",
    "    return [input[i:i+k] for i in range(len(input)-k+1)]\n",
    "\n",
    "# read in the genome\n",
    "with open('carsonella_ruddii_genome.txt') as f:\n",
    "    genome = f.read()\n",
    "\n",
    "# I will use reads of length 150 (k) and distance between reads of 1,000 bp (d)\n",
    "k = 150\n",
    "d = 1000\n",
    "\n",
    "# get all kmers\n",
    "reads = get_kmers(genome, k+d+k)\n",
    "\n",
    "# make into paired reads with a distance of 1000 between reads and separate with a \"|\"\n",
    "paired_reads = [(read[0:k] + \"|\" + read[k+d+k-k:k+d+k]) for read in reads]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the paired reads of the genome of *Carsonella ruddii*, we can make the de Bruijn graph from these."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the prefix for a read pair\n",
    "def paired_prefix(pair):\n",
    "    return (pair[0][:-1], pair[1][:-1])\n",
    "\n",
    "# get the suffix for a read pair\n",
    "def paired_suffix(pair):\n",
    "    return (pair[0][1:], pair[1][1:])\n",
    "\n",
    "# make the de Bruijn graph from the paired reads\n",
    "def de_bruijn_graph_from_read_pairs(read_pairs):\n",
    "    # make an empty graph\n",
    "    graph = {}\n",
    "    \n",
    "    # split the read_pairs at '|'\n",
    "    for pattern in read_pairs:\n",
    "        split = pattern.split(\"|\")\n",
    "        \n",
    "        suffix = paired_suffix(split)\n",
    "        prefix = paired_prefix(split)\n",
    "        \n",
    "        # add prefix:suffix pairs\n",
    "        if prefix in graph.keys():\n",
    "            graph[prefix].append(suffix)\n",
    "        else:\n",
    "            graph[prefix] = [suffix]\n",
    "    \n",
    "    \n",
    "    return graph\n",
    "\n",
    "graph = de_bruijn_graph_from_read_pairs(paired_reads)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the de Bruijn graph, we want to get the Eulerian path that passes each edge exactly once. In a perfectly closed graph, we can keep randomly walking across the graph while keeping track of unexplored nodes until no unexplored nodes are left. However, even in this genome with perfect coverage, the graph is not closed because the DNA sequence is not circular. Therefore, we can use the Eulerian_path algorithm to connect the start- and end-node with an edge and then find the Eulerian cycle that spells out the complete genome. This will take some minutes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "# walk a random path through a graph but check for unbalanced nodes\n",
    "def random_walk(graph, start):\n",
    "    # make a list of all nodes\n",
    "    node_list = [key for key in graph.keys()]\n",
    "    \n",
    "    # start the cycle at the starting node\n",
    "    cycle = [start]\n",
    "\n",
    "    # make a random cycle\n",
    "    while True:\n",
    "        # check if this is an unbalanced node by seeing if the node is in the node list\n",
    "        if cycle[-1] not in node_list:\n",
    "            break\n",
    "        # make a random choice for edges at each node\n",
    "        choice = random.choice(graph[cycle[-1]])\n",
    "        graph[cycle[-1]].remove(choice)\n",
    "        cycle.append(choice)\n",
    "        # when the cycle is complete break\n",
    "        if cycle[-1] == start:\n",
    "            break\n",
    "    \n",
    "    return cycle\n",
    "\n",
    "# Get the Eulerian cycle of an Eulerian directed graph\n",
    "def eulerian_cycle(graph, start):\n",
    "    # make a list of all nodes\n",
    "    node_list = [key for key in graph.keys()]\n",
    "\n",
    "    # choose a random starting node if no start node is given\n",
    "    if not isinstance(start, int) or isinstance(start, str):\n",
    "        start = random.choice(list(graph.keys()))\n",
    "\n",
    "    # walk a random cycle starting from the start node\n",
    "    cycle = random_walk(graph, start)\n",
    "\n",
    "    # keep track of the unexplored nodes\n",
    "    unexplored = [i for i in node_list if len(graph[i]) != 0]\n",
    "\n",
    "    # continue walking random cycles until no unexplored nodes are left\n",
    "    while len(unexplored) != 0:\n",
    "        # make a new graph with only the unexplored nodes\n",
    "        new_graph = {key: graph[key] for key in unexplored}\n",
    "        # choose a new random starting node from the list of unexplored nodes\n",
    "        new_start = random.choice([ele for ele in cycle if ele in unexplored])\n",
    "        # walk a new random cycle\n",
    "        new_cycle = random_walk(new_graph, new_start)\n",
    "        # first redo the previous cycle starting at the new start_point\n",
    "        index = cycle.index(new_start)\n",
    "        if index != 0:\n",
    "            old_cycle = cycle[index::] + cycle[1:index]\n",
    "        else:\n",
    "            old_cycle = cycle[:-1]\n",
    "        # the new cycle is the combination of the old cycle and the new cycle\n",
    "        cycle = old_cycle + new_cycle\n",
    "        # update the unexplored nodes\n",
    "        unexplored = [i for i in node_list if len(graph[i]) != 0]\n",
    "\n",
    "    # return the final Eulerian path\n",
    "    return cycle\n",
    "\n",
    "# Get the Eulerian path of an Eulerian directed graph\n",
    "def eulerian_path(graph):\n",
    "    # make a list of all nodes\n",
    "    node_list = [key for key in graph.keys()]\n",
    "    \n",
    "    # make a list of all edges\n",
    "    edge_list = []\n",
    "    for i in graph.values():\n",
    "        for value in i:\n",
    "            edge_list.append(value)\n",
    "\n",
    "    # add missing node if there is one\n",
    "    if len(edge_list) != len(node_list)+1:\n",
    "        difference = set(edge_list) - set(node_list)\n",
    "        graph.update({list(difference)[0]: []})\n",
    "\n",
    "    # find unbalanced nodes\n",
    "    unbalanced_nodes = [key for key, value in graph.items() if len(value) != edge_list.count(key)]\n",
    "\n",
    "    # determing which is the first and which is the last node\n",
    "    # outgoing number\n",
    "    outgoing_nodes = len(graph[unbalanced_nodes[0]])\n",
    "    # incoming number\n",
    "    incoming = []\n",
    "    for ele in graph.values():\n",
    "        for value in ele:\n",
    "            incoming.append(value)\n",
    "    # outgoing number\n",
    "    incoming_nodes = incoming.count(unbalanced_nodes[0])\n",
    "\n",
    "    if incoming_nodes < outgoing_nodes:\n",
    "        start_node = unbalanced_nodes[0]\n",
    "        end_node = unbalanced_nodes[1]\n",
    "    else:\n",
    "        start_node = unbalanced_nodes[1]\n",
    "        end_node = unbalanced_nodes[0]\n",
    "\n",
    "    # append the edge between end and start node or create if it does not yet exist\n",
    "    if end_node in graph:\n",
    "        graph[end_node].append(start_node)\n",
    "    else:\n",
    "        graph.update({end_node: [start_node]})\n",
    "    \n",
    "    # run eulerian_cycle\n",
    "    output = eulerian_cycle(graph, start = start_node)\n",
    "    \n",
    "    #return start_node[0]\n",
    "    # the path has to be reordered to start at the start_node and end at the end_node but\n",
    "    # only if the correct starting node is not already at the start!\n",
    "    if output[0] != start_node:\n",
    "        start_indexes = [index for index, ele in enumerate(output) if ele == start_node]\n",
    "        end_index = output.index(end_node)\n",
    "        start_position = [ele for ele in start_indexes if ele == end_index+1][0]\n",
    "        # return the reordered eulerian path\n",
    "        return output[start_position::]+output[1:start_position]\n",
    "    else:\n",
    "        return output[:-1]\n",
    "\n",
    "# walk to Eulerian path through the de Bruijn graph\n",
    "path = eulerian_path(graph)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the Eulerian path, the reads can be overlapped to generate the final assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert the tuple to a list readable by string_gapped_patterns\n",
    "input = []\n",
    "for ele in path:\n",
    "    input.append(\"|\".join(list(ele)))\n",
    "\n",
    "def string_spelled_by_patterns(patterns):\n",
    "    str = patterns[0]\n",
    "    for i in range(1, len(patterns)):\n",
    "        str += patterns[i][-1]\n",
    "    return str\n",
    "\n",
    "# this function overlaps strings with gapped patterns\n",
    "def string_gapped_patterns(patterns, k, d):\n",
    "    # first extract the first and last kmers between the gap\n",
    "    first_patterns = []\n",
    "    last_patterns = []\n",
    "    for pattern in patterns:\n",
    "        split = pattern.split(\"|\")\n",
    "        first_patterns.append(split[0])\n",
    "        last_patterns.append(split[1])\n",
    "\n",
    "    # reconstruct the first patterns\n",
    "    prefix_string = string_spelled_by_patterns(first_patterns)\n",
    "    # reconstruct the second patterns\n",
    "    suffix_string = string_spelled_by_patterns(last_patterns)\n",
    "\n",
    "    # check if they match and if so return the complete string\n",
    "    for i in range(k+d+1, len(prefix_string)):\n",
    "        if prefix_string[i] != suffix_string[i-k-d]:\n",
    "            return \"there is no string spelled by the gapped patterns\"\n",
    "    return prefix_string + suffix_string[len(suffix_string)-k-d:]\n",
    "\n",
    "assembly = string_gapped_patterns(input, k, d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of the genome of Carsonella ruddii is 166163 bp\n",
      "The length of the assembly is 166163 bp\n",
      "Is the assembly the same as the genome? True\n"
     ]
    }
   ],
   "source": [
    "print(f\"The length of the genome of Carsonella ruddii is {len(genome)} bp\")\n",
    "print(f\"The length of the assembly is {len(assembly)} bp\")\n",
    "\n",
    "print(f\"Is the assembly the same as the genome? {assembly == genome}\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although this example is quite a simple one with full coverage of error-free reads, it has shown me that genome assembly from short reads is a very difficult task. When sequencing errors are introduced to the reads, finding a Eulerian path suddenly becomes extremely non-trivial. Therefore, scaffolding to a reference genome is needed in order to try to correctly order contigs.\n",
    "\n",
    "This project did teach me a lot about the mechanism of genome assembly and it's pittfalls, which allows me to better utilize assembly software in the future. It also further taught me valuable skills on writing algorithms in Python."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
